crates/floresta-wire/src/p2p_wire/peer.rs

```diff
@@ -56,23 +58,163 @@ enum State {
-    stream: T,
+// Define the actor
+// Define the actor
+pub struct TcpStreamActor {
+    pub stream: TcpStream,
+    pub receiver: UnboundedReceiver<TcpStreamCommand>,
+    pub sender: UnboundedSender<ReaderMessage>,
+    pub network: Network,
+}
+
+impl TcpStreamActor {
+    pub async fn start(mut self) {
+        let (reader, mut writer) = tokio::io::split(self.stream);
+        let mut read_future = pin!(Self::read_loop_inner(reader));
+
+        loop {
+            futures::select! {
+                command = self.receiver.recv().fuse() => {
+                    let Some(command) = command else {
+                        return;
+                    };
+                    match command {
+                        TcpStreamCommand::Write(data) => {
+                            if let Err(e) = Self::handle_write(&mut writer, data).await {
+                                let _ = self.sender.send(ReaderMessage::Error(e));
+                                return;
+                            }
+                        }
+                        TcpStreamCommand::Shutdown => {
+                             if let Err(e) = Self::handle_shutdown(&mut writer).await {
+                                let _ = self.sender.send(ReaderMessage::Error(e));
+                            }
+                            return;
+                        }
+                    }
+                },
+                result = read_future.as_mut().fuse() => {
+                    match result {
+                        Ok(message) => {
+                            let _ = self.sender.send(ReaderMessage::Message(message));
+                        }
+                        Err(e) => {
+                            let _ = self.sender.send(ReaderMessage::Error(e));
+                            return;
+                        }
+                    }
+                },
+            }
+        }
+    }
+
+    async fn read_loop_inner(mut stream: ReadHalf<TcpStream>) -> Result<RawNetworkMessage> {
+        let mut data: Vec<u8> = vec![0; 24];
+
+        // Read the header first, so learn the payload size
+        stream.read_exact(&mut data).await?;
+        let header: P2PMessageHeader = deserialize_partial(&data)?.0;
+
+        // Network Message too big
+        if header.length > (1024 * 1024 * 32) as u32 {
+            return Err(PeerError::MessageTooBig);
+        }
+
+        data.resize(24 + header.length as usize, 0);
+
+        // Read everything else
+        stream.read_exact(&mut data[24..]).await?;
+
+        // Intercept block messages
+        if header._command[0..5] == [0x62, 0x6c, 0x6f, 0x63, 0x6b] {
+            let mut block_data = vec![0; header.length as usize];
+            block_data.copy_from_slice(&data[24..]);
+
+            let message: UtreexoBlock = deserialize(&block_data)?;
+            return Ok(RawNetworkMessage::new(
+                header.magic,
+                NetworkMessage::Block(message),
+            ));
```

Comment:
@Davidson-Souza sent a block message here to replace loop, should work ig?

Reply:
Also it was throwing error while using `UtreexoBlock` of floresta_wire, had to use `bitcoin::p2p::utreexo::UtreexoBlock` instead.

Reply:
> @Davidson-Souza sent a block message here to replace loop, should work ig?

Do you mean replace the loop in `read_loop_inner`? I fear if we get a message from peer through the channel while we are in the middle of a `read` this could drop the future and make us miss a package. We have to test it running to see what happens.

> Also it was throwing error while using UtreexoBlock of floresta_wire, had to use bitcoin::p2p::utreexo::UtreexoBlock instead.

Did you rebase with `master`? 'Cause we are phasing out this fork of `bitcoin`. So we can't use `bitcoin::p2p::utreexo` or the inners of `Target`